src/www/app/clipboard.ts

```diff
@@ -12,35 +12,18 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-/// <reference path='../../types/ambient/clipboard.d.ts'/>
+/// <reference path='../../types/outline.d.ts'/>
 
-export type ClipboardListener = (text: string) => void;
+import {Clipboard} from './clipboard_common';
+import {CordovaClipboard} from './clipboard_cordova';
+import {ElectronClipboard} from './clipboard_electron';
 
-export interface Clipboard {
-  // Returns the current contents of the clipboard.
-  getContents(): Promise<string>;
-
-  // Sets a callback to be invoked when the contents of the clipboard should be read. When this
-  // happens is implementation-dependent, e.g. it could be timer-based or, as is the case on mobile,
-  // whenever the app is brought to the foreground.
-  setListener(listener: ClipboardListener): void;
-}
-
-// Generic clipboard. Implementations should only have to implement getContents().
-export class AbstractClipboard implements Clipboard {
-  private listener: ClipboardListener|null = null;
-
-  getContents(): Promise<string> {
-    return Promise.reject(new Error('unimplemented skeleton method'));
-  }
-
-  setListener(listener: ClipboardListener) {
-    this.listener = listener;
-  }
-
-  emitEvent() {
-    if (this.listener) {
-      this.getContents().then(this.listener);
-    }
+export function getClipboard(): Clipboard {
+  if (outline.WEB_PLATFORM === 'cordova') {
+    return new CordovaClipboard();
```

Comment:
I believe you should be able to do this with dynamic `import()` expressions - https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html#:~:text=Dynamic%20import%20expressions%20are%20a,arbitrary%20point%20in%20your%20program.

Reply:
Love this btw, it's start of the HAL @jyyi1 and i talked about! I wonder if we should group these together e.g. like

```
/lib
  /clipboard
    cordova.ts
    electron.ts
    common.ts
    index.ts (entrypoint)
```

Reply:
just so if we have to move it we don't need to remember which files to take

Reply:
That structure is very similar to what Kotlin Multiplatform does: https://kotlinlang.org/docs/mpp-discover-project.html#source-sets.

Perhaps we need the concept of a "module" in our repository. It could be a local npm workspace, and they can depend on other local workspaces.

Unfortunately dynamic `import()` doesn't work, since it still tried to compile the electron code on Cordova and fails.

Reply:
Ah, i think you need to do something in the index at the top level like...

```
let CordovaClipboard;
  if (outline.WEB_PLATFORM === 'cordova') {




export default CordovaClipboard;
```

Reply:
the top level is run at compile time and should delegate iirc, rather than doing it in the function, where webpack can't assume what will be called at runtime

Reply:
we may need top-level await support in webpack

Reply:
> Perhaps we need the concept of a "module" in our repository. It could be a local npm workspace, and they can depend on other local workspaces.

many modern monorepo concepts have a top-level folder for binaries and a top-level folder for modules like you describe here

Reply:
hm, i do wonder if you can't just `require(`./path/to/${outline.WEB_PLATFORM}`);` - https://webpack.js.org/guides/dependency-management/#require-with-expression